File: validAnagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

class Solution(object):
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        
        dict_s = {}
        dict_t = {}
        for char in s:
            if char in dict_s:
                dict_s[char]+=1
            else:
                dict_s[char]=1

        for char in t:
            if char in dict_t:
                dict_t[char]+=1
            else:
                dict_t[char]=1
        
        if dict_s == dict_t:
            return True

# Explanation
# If the length of both strings are different, they cannot be an anagram
# Initialize two different dictionaries
# For each char in the string s, if that char is already in dict_s, we increment it by one, otherwise we assign it the value of 1
# Repeat same steps for string t
# If both dictionaries are equivalent, return TRUE

File: validAnagram.c
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.
// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

bool isAnagram(char *s, char *t) {
    if (strlen(s) != strlen(t)) {
        return false;
    }

    int len1 = strlen(s);
    int len2 = strlen(t);

    int letterCount1[26] = {0};
    int letterCount2[26] = {0};    

    for(int i = 0; i < len1; i++)
    {
        int lower = tolower(s[i]);
        letterCount1[lower - 'a']++;
    }

    for(int i = 0; i < len2; i++)
    {
        int lower = tolower(t[i]);
        letterCount2[lower - 'a']++;
    }

    for(int i = 0; i < 26; i++)
    {
        if (letterCount1[i] != letterCount2[i])
        {
            return false;
        }
    }

    return true;
}

// Explanation
// Check if strings are of equal length, otherwise return false
// Obtain length of both strings
// Initialize two arrays for each string with 26 indexes (for each letter in the alphabet) each starting at 0
// Loop over the length of the first string, convert the letter to lowercase, subtract it from 'a' so we get the apropriate index, and add 1 to the appropriate index
// Do the same for the second string
// Loop over the entirety of both arrays, if each index is equal return true, otherwise return false
